fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/PromoteValue.java

```diff
@@ -186,8 +188,7 @@ public static MessageHelpers.CoercionTrieNode computePromotionsTrie(@Nonnull fin
         }
 
         if (currentType.isPrimitive()) {
-            SemanticException.check(targetType.isPrimitive(), SemanticException.ErrorCode.INCOMPATIBLE_TYPE);
-            if (!isPromotionNeeded(currentType, targetType)) {
+            if (currentType.getTypeCode() != Type.TypeCode.NULL && !isPromotionNeeded(currentType, targetType)) {
```

Comment:
Please move condition `currentType.getTypeCode() != Type.TypeCodeNULL` to `isPromotionNeeded` and restore the check.

Reply:
Moved the condition, but cannot put back the condition without explicitly testing if the `currentType` in NULL. I think the semantic check here is a bit redundant because the coercion will promotion will anyways return null if not possible.

Reply:
semantic check also happens in `isPromotionNeeded` so indeed redundant.